### ISSUE TITLE
[question] deploy vs deployer

### ISSUE THREAD

**Issue by jsallay** [conan-io/conan#17143]
### What is your question?

I'm trying to understand deployment from the conan cache to final systems and I'm confused on the interaction of the `deploy` function in a conanfile and `deployer`s.  I have a conanfile with a custom deploy function that updates some of the files and I want to run a deployer such as `runtime_deploy`.

First off, why is calling the conanfile `deploy` function optional?  I have to specify which packages I want via the `deployer-package` argument.  I might be missing something but in my case, the function is required for the package to work post deployment.  So it is an unpleasant surprise to the user that it is required.

Second, it looks like the `deploy` function and the `runtime_deploy` run independently.  If I run `conan install --requires mypkg/1.0.0 --deployer-package=mypkg --deployer=runtime_deploy`.  It looks like it calls my `deploy` function and then later calls the `runtime_deployer` which overwrites the files from my deploy function.

I am confused as to how these should (or shouldn't) interact.  Is there a better way to accomplish what I am trying to do?

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by jsallay** [comment]
Thanks for explaining it, I think that makes sense.  I always call deploy with specific deploy arguments like `--deployer-folder`.  So I hadn't considered the case of installing without wanting to deploy vs deploying to the local folder.  There isn't a way to tell the user intention without a command line argument - hence the `--deployer-package`.

As far as the two pieces working together, I had assumed incorrectly that the `runtime_deploy` would call my conanfile's deploy method.  I can see why this wouldn't work with how things are currently setup, but it might be nice feature.

**Comment by memsharded** [comment]
> So I hadn't considered the case of installing without wanting to deploy vs deploying to the local folder. There isn't a way to tell the user intention without a command line argument - hence the --deployer-package.

Yes, this is by far the most common use case, installing packages in the Conan cache and using them from there, without deploying. Like 99% of the usage is this way, and deployment is only done by some users at the very last mile when they need to deploy the final artifacts by other means (creating an installer for the customer, or a .deb package to deploy in servers).

> As far as the two pieces working together, I had assumed incorrectly that the runtime_deploy would call my conanfile's deploy method. I can see why this wouldn't work with how things are currently setup, but it might be nice feature.

Yes, they are not really connected. I am not sure if connecting them would be that great, as there are many different corner cases, like for example what if calling several deployers like ``--deployer=runtime_deploy --deployer=license_deploy``, they could end calling twice the recipes ``deploy()`` with potential collisions.  These features aim to be quite orthogonal

**Comment by jsallay** [comment]
I think my confusion comes from not understanding that these are orthogonal features.  It is confusing since they both contain deploy.  I think it would be helpful to update the documentation to reflect this:  Something like:
`The deploy function is an independent feature from a deployer.  If a conan install command contains --deployer-package and --deployer arguments, the deploy functions of the selected package conanfiles will be run prior to the deployers.`

**Comment by jsallay** [comment]
Also, I understand the use case of installing into the cache without wanting to deploy anywhere.  That is what I use the majority of the time as well.  What I hadn't considered was deploying without using the `--deployer-folder` argument.  I had assumed that a different argument would always be present to indicate that we were deploying to a different location.

**Comment by memsharded** [comment]
Thanks for the feedback @jsallay, sorry that I didn't follow up earlier.
It seems that the docs could get some improvements, I am moving this ticket to the docs repo then, but please let me know if you have any further question or issue. Thanks!
